library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java

```diff
@@ -51,291 +66,607 @@
-
-  protected transient FSStorageAgent storageAgent;
+  private static final String WAL_FILE_NAME = "wal";
 
   /**
-   * Recovery path relative to app path where state is saved.
+   * Recovery filePath relative to app filePath where state is saved.
    */
   @NotNull
-  private String recoveryPath;
+  private String recoveryPath = DEF_RECOVERY_PATH;
 
   private boolean isRecoveryPathRelativeToAppPath = true;
 
   /**
-   * largest window for which there is recovery data across all physical operator instances.
+   * This is not null only for one physical instance.<br/>
+   * It consists of operator ids which have been deleted but have some state that can be replayed.
+   * Only one of the instances would be handling (modifying) the files that belong to this state. <br/>
+   * The value is assigned during partitioning.
    */
-  protected transient long largestRecoveryWindow;
+  private Set<Integer> deletedOperators;
+
+  private boolean repartitioned;
 
   /**
-   * This is not null only for one physical instance.<br/>
-   * It consists of operator ids which have been deleted but have some state that can be replayed.
-   * Only one of the instances would be handling (modifying) the files that belong to this state.
+   * Used when it is not necessary to replay every streaming/app window.
+   * Used by {@link IncrementalCheckpointManager}
    */
-  protected Set<Integer> deletedOperators;
+  private boolean relyOnCheckpoints;
 
   /**
-   * Sorted mapping from window id to all the operators that have state to replay for that window.
+   * largest window for which there is recovery data across all physical operator instances.
    */
-  protected final transient TreeMultimap<Long, Integer> replayState;
+  private transient long largestRecoveryWindow = Stateless.WINDOW_ID;
+
+  private final FSWindowReplayWAL wal = new FSWindowReplayWAL();
 
-  protected transient FileSystem fs;
-  protected transient Path appPath;
+  //operator id -> wals (sorted)
+  private final transient Map<Integer, FSWindowReplayWAL> readOnlyWals = new HashMap<>();
+
+  private transient String statePath;
+  private transient int operatorId;
+
+  private final transient Kryo kryo = new Kryo();
+
+  private transient FileContext fileContext;
 
   public FSWindowDataManager()
   {
-    replayState = TreeMultimap.create();
-    largestRecoveryWindow = Stateless.WINDOW_ID;
-    recoveryPath = DEF_RECOVERY_PATH;
+    kryo.setClassLoader(Thread.currentThread().getContextClassLoader());
   }
 
   @Override
   public void setup(Context.OperatorContext context)
   {
-    Configuration configuration = new Configuration();
+    operatorId = context.getId();
+
     if (isRecoveryPathRelativeToAppPath) {
-      appPath = new Path(context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + recoveryPath);
+      statePath = context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + recoveryPath;
     } else {
-      appPath = new Path(recoveryPath);
+      statePath = recoveryPath;
     }
```

Comment:
Yeah this is a problem. By default `statePath` will be shared by all logical operator. There is another ticket APEXMALHAR-2068 opened for that. Also the fix will need the recent change to Apex Core (APEXCORE-448 ) which will not be there until the core version in malhar changes.